File: crates/taflow-core/src/stream/decay_linear.rs
//! Persistent WorldQuant Alpha101 linear-decay moving average state.

use super::{StreamingIndicator, WeightedMovingAverage};
use crate::error::TaResult;

/// Linear-decay weighted moving average with a persistent bounded state.
#[derive(Debug, Clone)]
pub struct DecayLinear {
    inner: WeightedMovingAverage,
}

impl DecayLinear {
    /// Construct a validated linear-decay state.
    pub fn new(timeperiod: usize) -> TaResult<Self> {
        Ok(Self {
            inner: WeightedMovingAverage::new(timeperiod)?,
        })
    }

    /// Extend the state with chronological values and aligned warm-up output.
    pub fn extend_slice_into(&mut self, input: &[f64], output: &mut Vec<f64>) {
        output.extend(
            input
                .iter()
                .copied()
                .map(|value| self.append(value).unwrap_or(f64::NAN)),
        );
    }
}

impl StreamingIndicator for DecayLinear {
    type Output = f64;

    fn append(&mut self, input: f64) -> Option<f64> {
        self.inner.append(input)
    }

    fn value(&self) -> Option<f64> {
        self.inner.value()
    }

    fn reset(&mut self) {
        self.inner.reset();
    }
}

File: python/taflow/indicators/decay_linear.py
"""Canonical native-backed linear-decay adapter."""

from typing import Any
import numpy as np
from .._adapter_protocol import adapter_length
from .._native import WeightedMovingAverage as _Native
from .._series import as_float64_series


class DecayLinear:
    """WorldQuant linear-decay weighted moving average."""

    def __init__(self, input: Any, timeperiod: int = 30) -> None:
        self._state = _Native(int(timeperiod))
        self.extend(input)

    def append(self, input: float) -> "DecayLinear":
        self._state.append(float(input))
        return self

    def extend(self, input: Any) -> "DecayLinear":
        values = as_float64_series(input)
        self._state.extend(values)
        return self

    def compute(self) -> np.ndarray:
        return self._state.compute()

    @property
    def value(self) -> float | None:
        return self._state.value

    def reset(self) -> "DecayLinear":
        self._state.reset()
        return self

    def __len__(self) -> int:
        return adapter_length(self)
